packages/blitz-next/src/index-server.ts

```diff
@@ -1,3 +1,72 @@
+import { GetServerSideProps, NextApiHandler, NextApiRequest, NextApiResponse } from "next";
+
 export * from "./index-browser"
 
-export const todoServer = true
+type TemporaryAny = any
+
+export interface DefaultCtx { }
+export interface Ctx extends DefaultCtx { }
+
+export interface MiddlewareResponse<C = Ctx> extends NextApiResponse {
+  blitzCtx: C
+  blitzResult: unknown
+}
+
+export type BlitzMiddleware = (handler: NextApiHandler) => (req: NextApiRequest, res: MiddlewareResponse, ctx?: TemporaryAny) => Promise<void>;
+
+// todo: better naming
+export type Middleware = (req: NextApiRequest, res: MiddlewareResponse, next: (error?: Error) => Promise<void> | void) => Promise<void> | void
+
+export type BlitzPlugin = {
+  middlewares: Middleware[];
+  // particular plugin creators should return unified BlitzPlugin type, so that this file doesn't need to know about particular plugins
+}
+
+
+const runMiddlewares = async (middlewares: Middleware[], req: NextApiRequest, res: MiddlewareResponse) => {
+  const promises = middlewares.reduce((acc, middleware) => {
+    const promise = new Promise((resolve, reject) => {
+      middleware(req, res, result =>
+        result instanceof Error ? reject(result) : resolve(result),
+      );
+    });
+    return [...acc, promise];
+  }, [] as Promise<unknown>[]);
+
+  await Promise.all(promises);
+}
+
```

Comment:
We need to somehow support error middleware for things like capturing to Sentry:

```ts
async (req, res, next) => {
  try {
    await next()
  } catch (error) {
    Sentry.captureException(error)
    throw error
  }
}
```

I'm not sure if legacy blitz middleware supported the above or not, but for sure it didn't work for RPC errors, because the old RPC middleware would catch the error and turn it into a 200 response.

So we may need to add an extra error middleware type or event type so we can capture RPC resolver errors.